### DISCUSSION TITLE
Linked code made not entrant

### DISCUSSION THREAD

**Discussion by TomerYogev** [apangin/nalim#13] [Q&A] [answered]
I tried using `nalim` to link native code to a native Java function. It sometimes works but in certain flows the JIT decides to make it not entrant, causing an `UnsatisfiedLinkError`. Any ideas?
(I'm running on JDK11)

**Comment by apangin** [comment]
Do you have an example that demonstrates the issue?

**Reply by TomerYogev** [reply]
Managed to boil it down to this: I'm attaching a Java agent with an `agentmain` that does:
1. Link native code to function.
2. `instrumentation.retransformClasses(String.class);`
3. Call the linked function.

This throws an `UnsatisfiedLinkError` for me. It doesn't seem to matter which class I retransform, although it _doesn't_ happen if I do it from `premain`! It doesn't reproduce on JDK19+.

**Reply by apangin** [reply, marked as answer]
Thank you. This seems to be related to the flushing of the enitre CodeCache on the first class redefinition.
This has been fixed in [JDK-8324241](https://bugs.openjdk.org/browse/JDK-8324241).

If the assumption is correct, then claiming `Can-Redefine-Classes` capability in nalim java agent should solve the issue.

**Reply by TomerYogev** [reply]
Very interesting! Regarding `Can-Redefine-Classes`, do you mean defining it for the agent that is loaded when the process starts? Is there a way to fix this for when the process starts without it and I need to attach it dynamically? Other than manually retransforming some class before linking the native code?

**Reply by apangin** [reply]
> do you mean defining it for the agent that is loaded when the process starts?

Yes.

> Is there a way to fix this for when the process starts without it and I need to attach it dynamically? Other than manually retransforming some class before linking the native code?

The agent can do this automatically on the first attach. I'm not aware of other workarounds right now; this requires deeper research.
